workflow/controller/steps.go

```diff
@@ -330,7 +330,10 @@ func (woc *wfOperationCtx) executeStepGroup(ctx context.Context, stepGroup []wfv
 			completed = false
 		} else if childNode.Completed() {
 			hasOnExitNode, onExitNode, err := woc.runOnExitNode(ctx, step.GetExitHook(woc.execWf.Spec.Arguments), childNode, stepsCtx.boundaryID, stepsCtx.tmplCtx, "steps."+step.Name, stepsCtx.scope)
-			if hasOnExitNode && (onExitNode == nil || !onExitNode.Fulfilled() || err != nil) {
+			if err != nil {
+				return node, err
+			}
+			if hasOnExitNode && (onExitNode == nil || !onExitNode.Fulfilled()) {
 				// The onExit node is either not complete or has errored out, return.
```

Comment:
It would be best to update the explanation in the comments as well.

Reply:
done